### ISSUE TITLE
ConceptAprilTagLocalization doesn't work

### ISSUE THREAD

**Issue by ikarib** [deltacv/EOCV-Sim#116]
**Is your feature request related to a problem? Please describe.**
Please add ability to use example code from ConceptAprilTagLocalization. It fails due to missing functions in AprilTagProcessorImpl.java. I think you removed computeRobotPose due to missing OpenGLMatrix implementation.

**Describe the solution you'd like**
I would like you to add android/opengl/Matrix.java from [Android SDK sources](https://github.com/AndroidSDKSources/android-sdk-sources-for-api-level-29/raw/refs/heads/master/android/opengl/Matrix.java) and replace the following Java native calls multiplyMM and multiplyMV:

    public static void multiplyMM(float[] result, int resultOffset,
            float[] lhs, int lhsOffset, float[] rhs, int rhsOffset) {
        for (int i=0 ; i<4 ; i++) {
            float rhs_i0 = rhs[ rhsOffset + 4*i ];
            float ri0 = lhs[ lhsOffset ] * rhs_i0;
            float ri1 = lhs[ lhsOffset + 1 ] * rhs_i0;
            float ri2 = lhs[ lhsOffset + 2 ] * rhs_i0;
            float ri3 = lhs[ lhsOffset + 3 ] * rhs_i0;
            for (int j=1 ; j<4 ; j++) {
                float rhs_ij = rhs[ rhsOffset + 4*i + j ];
                ri0 += lhs[ lhsOffset + 4*j ] * rhs_ij;
                ri1 += lhs[ lhsOffset + 4*j + 1 ] * rhs_ij;
                ri2 += lhs[ lhsOffset + 4*j + 2 ] * rhs_ij;
                ri3 += lhs[ lhsOffset + 4*j + 3 ] * rhs_ij;
            }
            result[ resultOffset + 4*i ] = ri0;
            result[ resultOffset + 4*i+1 ] = ri1;
            result[ resultOffset + 4*i+2 ] = ri2;
            result[ resultOffset + 4*i+3 ] = ri3;
        }
    }

    public static void multiplyMV(float[] resultVec,
            int resultVecOffset, float[] lhsMat, int lhsMatOffset,
            float[] rhsVec, int rhsVecOffset) {
        for (int i=0 ; i<4 ; i++) {
            float rhs_i0 = rhsVec[ rhsVecOffset + 4*i ];
            float ri0 = lhsMat[ lhsMatOffset ] * rhs_i0;
            float ri1 = lhsMat[ lhsMatOffset + 1 ] * rhs_i0;
            float ri2 = lhsMat[ lhsMatOffset + 2 ] * rhs_i0;
            float ri3 = lhsMat[ lhsMatOffset + 3 ] * rhs_i0;
            for (int j=1 ; j<4 ; j++) {
                float rhs_ij = rhsVec[ rhsVecOffset + 4*i + j ];
                ri0 += lhsMat[ lhsMatOffset + 4*j ] * rhs_ij;
                ri1 += lhsMat[ lhsMatOffset + 4*j + 1 ] * rhs_ij;
                ri2 += lhsMat[ lhsMatOffset + 4*j + 2 ] * rhs_ij;
                ri3 += lhsMat[ lhsMatOffset + 4*j + 3 ] * rhs_ij;
            }
            resultVec[ resultVecOffset + 4*i ] = ri0;
            resultVec[ resultVecOffset + 4*i + 1 ] = ri1;
            resultVec[ resultVecOffset + 4*i + 2 ] = ri2;
            resultVec[ resultVecOffset + 4*i + 3 ] = ri3;
        }
    }

**Comment by serivesmejia** [comment]
Thank you for the issue report, I'll be adding the missing APIs on the next release of EOCV-Sim, ETA for next week as I try to sort out other issues as well..

**Comment by serivesmejia** [comment]
You can try out https://github.com/deltacv/EOCV-Sim/releases/tag/Dev for the implementation of the APIs you requested (along with the default workspace already containing ConceptAprilTagLocalization). However, due to some weird issue I'm trying to track, text isn't getting rendered into the viewport, everything else is functional though.

**Comment by ikarib** [comment]
Also for ConceptVisionColorSensor to work, please add android/graphics/Color.java from [Android SDK sources](https://raw.githubusercontent.com/AndroidSDKSources/android-sdk-sources-for-api-level-29/refs/heads/master/android/graphics/Color.java) and replace the following Java native calls nativeRGBtoHSV and nativeHSVtoColor with:
```
    private static void nativeRGBToHSV(int red, int green, int blue, float hsv[]) {        
        int min = Math.min(red, Math.min(green, blue));
        int max = Math.max(red, Math.max(green, blue));
        int delta = max - min;
    
        float v = (float) max / 255;
        if (v < 0 || v > 1.0f)
            throw new RuntimeException("RGB max is out of range");
    
        if (delta == 0) { // we're a shade of gray
            hsv[0] = 0;
            hsv[1] = 0;
            hsv[2] = v;
            return;
        }
        
        float s = (float) delta / max;
        if (s < 0 || s > 1.0f)
            throw new RuntimeException("RGB delta is out of range");
    
        float h;    
        if (red == max)
            h = (float) (green - blue) / delta;
        else if (green == max)
            h = 2.0f + (float) (blue - red) / delta;
        else // blue == max
            h = 4.0f + (float) (red - green) / delta;
        
        h *= 60;
        if (h < 0)
            h += 360.0f;

        if (h < 0 || h >= 360.0f)
            throw new RuntimeException("h is out of range");
        
        hsv[0] = h;
        hsv[1] = s;
        hsv[2] = v;
    }
        
    private static int nativeHSVToColor(int alpha, float hsv[]) {
        int s = hsv[1] < 0 ? 0 : hsv[1] >= 1.0f ? 255 : ((int) (hsv[1] * (1 << 16))) >> 8;
        int v = hsv[2] < 0 ? 0 : hsv[2] >= 1.0f ? 255 : ((int) (hsv[2] * (1 << 16))) >> 8;

        if (s == 0) // shade of gray
            return (alpha << 24) | (v << 16) | (v << 8) | v;
        int hx = (hsv[0] < 0 || hsv[0] >= 360.0f) ? 0 : ((int) (hsv[0]/60 * (1 << 16)));
        int f = hx & 0xFFFF;

        int v_scale = v+1;
        int p = ((255 - s) * v_scale) >> 8;
        int q = ((255 - (s * f >> 16)) * v_scale) >> 8;
        int t = ((255 - (s * (1 << 16 - f) >> 16)) * v_scale) >> 8;

        int r, g, b;

        if (hx >> 16 >= 6)
            throw new RuntimeException("hx is out of range");
        switch (hx >> 16) {
            case 0: r = v; g = t; b = p; break;
            case 1: r = q; g = v; b = p; break;
            case 2: r = p; g = v; b = t; break;
            case 3: r = p; g = q; b = v; break;
            case 4: r = t;  g = p; b = v; break;
            default: r = v; g = p; b = q; break;
        }
        return (alpha << 24) | (r << 16) | (g << 8) | b;
    }

```

**Comment by serivesmejia** [comment]
You can try the latest dev release with the requested changes now!

**Comment by serivesmejia** [comment]
Fixed in [v3.7.1](https://github.com/deltacv/EOCV-Sim/releases/tag/v3.7.1). Please open a new issue if you need more missing APIs
